scrapy/spidermiddlewares/depth.py

```diff
@@ -54,5 +56,11 @@ def _filter(request):
             response.meta['depth'] = 0
             if self.verbose_stats:
                 self.stats.inc_value('request_depth_count/0', spider=spider)
-
-        return (r for r in result or () if _filter(r))
+        result = list_to_simplevalue(result)
+        if isinstance(result, GeneratorType):
```

Comment:
It'd be much better if the feature doesn't require modifications in existing middlewares; there must be a way to support it in a framework. Otherwise users will have to modify all their custom middlewares in a similar way.